src/buffer.rs

```diff
@@ -189,3 +511,426 @@ fn move_and_insert_cjk() {
+    buf.insert_chars_at_cursor("word".to_string());
+    let pos3 = buf.char_pos();
+    buf.move_start();
+
+    buf.move_word();
+    assert_eq!(buf.char_pos(), pos1);
+    buf.move_word();
+    assert_eq!(buf.char_pos(), pos2);
+
+    buf.move_start();
+    assert!(buf.move_word_ws());
+    assert_eq!(buf.char_pos(), pos1);
+    assert!(!buf.move_word_ws());
+    assert_eq!(buf.char_pos(), pos3);
+}
+
+#[test]
+fn move_word_and_back() {
+    let mut buf = Buffer::new();
+    buf.insert_chars_at_cursor("here ".to_string());
+    let pos1 = buf.char_pos();
+    buf.insert_chars_at_cursor("are ".to_string());
+    let pos2 = buf.char_pos();
+    buf.insert_chars_at_cursor("some".to_string());
+    let pos3 = buf.char_pos();
+    buf.insert_chars_at_cursor("... ".to_string());
+    let pos4 = buf.char_pos();
+    buf.insert_chars_at_cursor("words".to_string());
+    let pos5 = buf.char_pos();
+
+    // make sure move_word() and move_word_back() are reflections of eachother
+
+    buf.move_start();
+    buf.move_word();
+    assert_eq!(buf.char_pos(), pos1);
+    buf.move_word();
+    assert_eq!(buf.char_pos(), pos2);
+    buf.move_word();
+    assert_eq!(buf.char_pos(), pos3);
+    buf.move_word();
+    assert_eq!(buf.char_pos(), pos4);
+    buf.move_word();
+    assert_eq!(buf.char_pos(), pos5);
+
+    buf.move_word_back();
+    assert_eq!(buf.char_pos(), pos4);
+    buf.move_word_back();
+    assert_eq!(buf.char_pos(), pos3);
+    buf.move_word_back();
+    assert_eq!(buf.char_pos(), pos2);
+    buf.move_word_back();
+    assert_eq!(buf.char_pos(), pos1);
+    buf.move_word_back();
+    assert_eq!(buf.char_pos(), 0);
+
+    buf.move_start();
+    buf.move_word_ws();
+    assert_eq!(buf.char_pos(), pos1);
+    buf.move_word_ws();
+    assert_eq!(buf.char_pos(), pos2);
+    buf.move_word_ws();
+    assert_eq!(buf.char_pos(), pos4);
+    buf.move_word_ws();
+    assert_eq!(buf.char_pos(), pos5);
+
+    buf.move_word_ws_back();
+    assert_eq!(buf.char_pos(), pos4);
+    buf.move_word_ws_back();
+    assert_eq!(buf.char_pos(), pos2);
+    buf.move_word_ws_back();
+    assert_eq!(buf.char_pos(), pos1);
+    buf.move_word_ws_back();
+    assert_eq!(buf.char_pos(), 0);
+}
+
+#[test]
+fn move_to_char() {
+    let mut buf = Buffer::new();
+    buf.insert_chars_at_cursor("words".to_string());
+    let pos1 = buf.char_pos();
+    buf.insert_chars_at_cursor(" wor".to_string());
+    let d_pos = buf.char_pos();
+    buf.insert_chars_at_cursor("ds".to_string());
+
+    buf.move_start();
+    assert!(buf.move_to_char_right(' ', 1));
+    assert_eq!(buf.char_pos(), pos1);
+    buf.move_end();
+    assert!(buf.move_to_char_left(' ', 1));
+    assert_eq!(buf.char_pos(), pos1);
+    buf.move_start();
+    assert_eq!(buf.move_to_char_right('z', 1), false);
+    assert_eq!(buf.char_pos(), 0);
+    buf.move_start();
+    assert!(buf.move_to_char_right('d', 2));
+    assert_eq!(buf.char_pos(), d_pos);
+}
+
+#[test]
+fn move_and_delete() {
```

Comment:
Smaller tests representing a single scenario to test make it easier to understand and gain insight into what broke when things change. Also, cargo runs tests in parallel, afaik, so multiple smaller tests will run faster that a single serial one.